File: src/main/java/com/danielremsburg/jaffolding/ui/ChartComponent.java
package com.danielremsburg.jaffolding.ui;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

import org.teavm.jso.JSBody;
import org.teavm.jso.JSObject;
import org.teavm.jso.JSProperty;
import org.teavm.jso.dom.html.HTMLCanvasElement;
import org.teavm.jso.dom.html.HTMLDocument;
import org.teavm.jso.dom.html.HTMLElement;
import org.teavm.jso.json.JSON;

import com.danielremsburg.jaffolding.Component;
import com.danielremsburg.jaffolding.State;

/**
 * A chart component that integrates with Chart.js.
 */
public class ChartComponent extends Component {
    private String type = "bar";
    private List<String> labels = new ArrayList<>();
    private List<Dataset> datasets = new ArrayList<>();
    private Map<String, Object> options = new HashMap<>();
    private Chart chart;
    private State<List<Map<String, Object>>> dataState;
    private String labelField;
    private String valueField;
    private String categoryField;
    private Function<Map<String, Object>, String> labelExtractor;
    private Function<Map<String, Object>, Number> valueExtractor;
    private Function<Map<String, Object>, String> categoryExtractor;
    
    public ChartComponent() {
        super("canvas");
        initializeStyles();
    }

    private void initializeStyles() {
        setStyle("width", "100%")
            .setStyle("height", "300px")
            .setStyle("max-height", "100%");
    }

    public ChartComponent setType(String type) {
        this.type = type;
        return this;
    }

    public ChartComponent setLabels(List<String> labels) {
        this.labels = new ArrayList<>(labels);
        return this;
    }

    public ChartComponent addDataset(String label, List<Number> data, String backgroundColor, String borderColor) {
        Dataset dataset = new Dataset();
        dataset.label = label;
        dataset.data = new ArrayList<>(data);
        dataset.backgroundColor = backgroundColor;
        dataset.borderColor = borderColor;
        dataset.borderWidth = 1;
        
        datasets.add(dataset);
        return this;
    }

    public ChartComponent setOptions(Map<String, Object> options) {
        this.options = new HashMap<>(options);
        return this;
    }

    public ChartComponent bindToDataState(State<List<Map<String, Object>>> dataState, 
                                         String labelField, 
                                         String valueField) {
        this.dataState = dataState;
        this.labelField = labelField;
        this.valueField = valueField;
        
        // Default extractors
        this.labelExtractor = data -> data.get(labelField).toString();
        this.valueExtractor = data -> (Number) data.get(valueField);
        
        // Subscribe to data changes
        dataState.subscribe(this::updateChartFromData);
        
        return this;
    }

    public ChartComponent bindToDataState(State<List<Map<String, Object>>> dataState, 
                                         String labelField, 
                                         String valueField,
                                         String categoryField) {
        bindToDataState(dataState, labelField, valueField);
        this.categoryField = categoryField;
        this.categoryExtractor = data -> data.get(categoryField).toString();
        
        return this;
    }

    public ChartComponent setLabelExtractor(Function<Map<String, Object>, String> extractor) {
        this.labelExtractor = extractor;
        return this;
    }

    public ChartComponent setValueExtractor(Function<Map<String, Object>, Number> extractor) {
        this.valueExtractor = extractor;
        return this;
    }

    public ChartComponent setCategoryExtractor(Function<Map<String, Object>, String> extractor) {
        this.categoryExtractor = extractor;
        return this;
    }

    private void updateChartFromData(List<Map<String, Object>> data) {
        if (data == null || data.isEmpty()) {
            return;
        }
        
        // Extract labels and organize data
        List<String> newLabels = new ArrayList<>();
        Map<String, Map<String, Number>> categoryData = new HashMap<>();
        
        if (categoryField != null) {
            // For grouped data (multiple datasets)
            for (Map<String, Object> item : data) {
                String label = labelExtractor.apply(item);
                String category = categoryExtractor.apply(item);
                Number value = valueExtractor.apply(item);
                
                if (!newLabels.contains(label)) {
                    newLabels.add(label);
                }
                
                categoryData.computeIfAbsent(category, k -> new HashMap<>())
                           .put(label, value);
            }
            
            // Clear existing datasets
            datasets.clear();
            
            // Create a dataset for each category
            String[] colors = {
                "#4285f4", "#ea4335", "#fbbc05", "#34a853", 
                "#673ab7", "#3f51b5", "#2196f3", "#03a9f4", 
                "#00bcd4", "#009688", "#4caf50", "#8bc34a"
            };
            
            int colorIndex = 0;
            for (Map.Entry<String, Map<String, Number>> entry : categoryData.entrySet()) {
                String category = entry.getKey();
                Map<String, Number> values = entry.getValue();
                
                List<Number> dataPoints = new ArrayList<>();
                for (String label : newLabels) {
                    dataPoints.add(values.getOrDefault(label, 0));
                }
                
                String color = colors[colorIndex % colors.length];
                addDataset(category, dataPoints, color + "33", color);
                colorIndex++;
            }
        } else {
            // For simple data (single dataset)
            List<Number> values = new ArrayList<>();
            
            for (Map<String, Object> item : data) {
                String label = labelExtractor.apply(item);
                Number value = valueExtractor.apply(item);
                
                newLabels.add(label);
                values.add(value);
            }
            
            // Clear existing datasets
            datasets.clear();
            
            // Add a single dataset
            List<String> backgroundColors = new ArrayList<>();
            List<String> borderColors = new ArrayList<>();
            
            for (int i = 0; i < values.size(); i++) {
                int hue = (i * 30) % 360;
                backgroundColors.add("hsla(" + hue + ", 70%, 60%, 0.2)");
                borderColors.add("hsla(" + hue + ", 70%, 60%, 1)");
            }
            
            Dataset dataset = new Dataset();
            dataset.label = valueField;
            dataset.data = values;
            dataset.backgroundColor = backgroundColors;
            dataset.borderColor = borderColors;
            dataset.borderWidth = 1;
            
            datasets.add(dataset);
        }
        
        // Update labels
        this.labels = newLabels;
        
        // Update the chart
        if (chart != null) {
            updateChart();
        }
    }

    private void createChart() {
        if (getElement() == null) {
            return;
        }
        
        HTMLCanvasElement canvas = (HTMLCanvasElement) getElement();
        
        // Prepare chart configuration
        JSObject config = createEmptyObject();
        setObjectProperty(config, "type", type);
        
        // Prepare data
        JSObject data = createEmptyObject();
        setObjectProperty(data, "labels", convertToJSArray(labels));
        
        // Prepare datasets
        JSObject[] jsDatasets = new JSObject[datasets.size()];
        for (int i = 0; i < datasets.size(); i++) {
            Dataset dataset = datasets.get(i);
            JSObject jsDataset = createEmptyObject();
            
            setObjectProperty(jsDataset, "label", dataset.label);
            setObjectProperty(jsDataset, "data", convertToJSArray(dataset.data));
            
            if (dataset.backgroundColor instanceof List) {
                setObjectProperty(jsDataset, "backgroundColor", convertToJSArray((List<?>) dataset.backgroundColor));
            } else {
                setObjectProperty(jsDataset, "backgroundColor", dataset.backgroundColor);
            }
            
            if (dataset.borderColor instanceof List) {
                setObjectProperty(jsDataset, "borderColor", convertToJSArray((List<?>) dataset.borderColor));
            } else {
                setObjectProperty(jsDataset, "borderColor", dataset.borderColor);
            }
            
            setObjectProperty(jsDataset, "borderWidth", dataset.borderWidth);
            
            jsDatasets[i] = jsDataset;
        }
        
        setObjectProperty(data, "datasets", convertToJSArray(jsDatasets));
        setObjectProperty(config, "data", data);
        
        // Prepare options
        JSObject jsOptions = createEmptyObject();
        setObjectProperty(jsOptions, "responsive", true);
        setObjectProperty(jsOptions, "maintainAspectRatio", false);
        
        // Add custom options
        for (Map.Entry<String, Object> entry : options.entrySet()) {
            setObjectProperty(jsOptions, entry.getKey(), entry.getValue());
        }
        
        setObjectProperty(config, "options", jsOptions);
        
        // Create the chart
        chart = createChart(canvas, config);
    }

    private void updateChart() {
        if (chart == null || getElement() == null) {
            return;
        }
        
        // Update labels
        JSObject data = getChartData(chart);
        setObjectProperty(data, "labels", convertToJSArray(labels));
        
        // Update datasets
        JSObject[] jsDatasets = new JSObject[datasets.size()];
        for (int i = 0; i < datasets.size(); i++) {
            Dataset dataset = datasets.get(i);
            JSObject jsDataset = createEmptyObject();
            
            setObjectProperty(jsDataset, "label", dataset.label);
            setObjectProperty(jsDataset, "data", convertToJSArray(dataset.data));
            
            if (dataset.backgroundColor instanceof List) {
                setObjectProperty(jsDataset, "backgroundColor", convertToJSArray((List<?>) dataset.backgroundColor));
            } else {
                setObjectProperty(jsDataset, "backgroundColor", dataset.backgroundColor);
            }
            
            if (dataset.borderColor instanceof List) {
                setObjectProperty(jsDataset, "borderColor", convertToJSArray((List<?>) dataset.borderColor));
            } else {
                setObjectProperty(jsDataset, "borderColor", dataset.borderColor);
            }
            
            setObjectProperty(jsDataset, "borderWidth", dataset.borderWidth);
            
            jsDatasets[i] = jsDataset;
        }
        
        setObjectProperty(data, "datasets", convertToJSArray(jsDatasets));
        
        // Update the chart
        updateChartData(chart);
    }

    @Override
    public HTMLElement render(HTMLElement parent) {
        HTMLElement element = super.render(parent);
        
        // Initialize the chart when the canvas is rendered
        createChart();
        
        return element;
    }

    /**
     * Dataset class for Chart.js.
     */
    public static class Dataset {
        public String label;
        public List<Number> data;
        public Object backgroundColor; // Can be String or List<String>
        public Object borderColor; // Can be String or List<String>
        public int borderWidth;
    }

    // JavaScript interop methods
    
    @JSBody(params = {}, script = "return {};")
    private static native JSObject createEmptyObject();
    
    @JSBody(params = {"obj", "key", "value"}, script = "obj[key] = value;")
    private static native void setObjectProperty(JSObject obj, String key, Object value);
    
    @JSBody(params = {"items"}, script = "return Array.from(items);")
    private static native JSObject convertToJSArray(Object[] items);
    
    @JSBody(params = {"items"}, script = "return Array.from(items);")
    private static native JSObject convertToJSArray(List<?> items);
    
    @JSBody(params = {"canvas", "config"}, script = "return new Chart(canvas, config);")
    private static native Chart createChart(HTMLCanvasElement canvas, JSObject config);
    
    @JSBody(params = {"chart"}, script = "return chart.data;")
    private static native JSObject getChartData(Chart chart);
    
    @JSBody(params = {"chart"}, script = "chart.update();")
    private static native void updateChartData(Chart chart);
    
    /**
     * Chart.js Chart interface.
     */
    private interface Chart extends JSObject {
        void update();
        void destroy();
    }
}

File: src/main/resources/static/js/jaffolding/ui/ChartComponent.js
/**
 * A chart component that integrates with Chart.js.
 * JavaScript implementation that mirrors the Java API.
 */
(function() {
  class ChartComponent extends jaffolding.Component {
    /**
     * Creates a new ChartComponent instance.
     */
    constructor() {
      super('canvas');
      this.type = 'bar';
      this.labels = [];
      this.datasets = [];
      this.options = {};
      this.chart = null;
      this.dataState = null;
      this.labelField = null;
      this.valueField = null;
      this.categoryField = null;
      this.labelExtractor = null;
      this.valueExtractor = null;
      this.categoryExtractor = null;
      
      this.initializeStyles();
    }

    /**
     * Initializes the chart styles.
     * @private
     */
    initializeStyles() {
      this.setStyle('width', '100%')
          .setStyle('height', '100%');
    }

    /**
     * Sets the chart type.
     * @param {string} type - The chart type (e.g., 'bar', 'line', 'pie')
     * @returns {ChartComponent} - The chart component instance for chaining
     */
    setType(type) {
      this.type = type;
      if (this.chart) {
        this.chart.destroy();
        this.createChart();
      }
      return this;
    }

    /**
     * Sets the chart labels.
     * @param {string[]} labels - The chart labels
     * @returns {ChartComponent} - The chart component instance for chaining
     */
    setLabels(labels) {
      this.labels = [...labels];
      if (this.chart) {
        this.updateChart();
      }
      return this;
    }

    /**
     * Adds a dataset to the chart.
     * @param {string} label - The dataset label
     * @param {number[]} data - The dataset values
     * @param {string} backgroundColor - The background color
     * @param {string} borderColor - The border color
     * @returns {ChartComponent} - The chart component instance for chaining
     */
    addDataset(label, data, backgroundColor, borderColor) {
      this.datasets.push({
        label,
        data: [...data],
        backgroundColor,
        borderColor,
        borderWidth: 1
      });
      
      if (this.chart) {
        this.updateChart();
      }
      
      return this;
    }

    /**
     * Sets the chart options.
     * @param {Object} options - The chart options
     * @returns {ChartComponent} - The chart component instance for chaining
     */
    setOptions(options) {
      this.options = {...options};
      if (this.chart) {
        this.updateChart();
      }
      return this;
    }

    /**
     * Binds the chart to a data state.
     * @param {State} dataState - The data state
     * @param {string} labelField - The field to use for labels
     * @param {string} valueField - The field to use for values
     * @returns {ChartComponent} - The chart component instance for chaining
     */
    bindToDataState(dataState, labelField, valueField) {
      this.dataState = dataState;
      this.labelField = labelField;
      this.valueField = valueField;
      
      // Default extractors
      this.labelExtractor = data => data[labelField];
      this.valueExtractor = data => data[valueField];
      
      // Subscribe to data changes
      dataState.subscribe(data => this.updateChartFromData(data));
      
      return this;
    }

    /**
     * Binds the chart to a data state with category grouping.
     * @param {State} dataState - The data state
     * @param {string} labelField - The field to use for labels
     * @param {string} valueField - The field to use for values
     * @param {string} categoryField - The field to use for categories
     * @returns {ChartComponent} - The chart component instance for chaining
     */
    bindToDataStateWithCategory(dataState, labelField, valueField, categoryField) {
      this.bindToDataState(dataState, labelField, valueField);
      this.categoryField = categoryField;
      this.categoryExtractor = data => data[categoryField];
      
      return this;
    }

    /**
     * Sets the label extractor function.
     * @param {Function} extractor - The label extractor function
     * @returns {ChartComponent} - The chart component instance for chaining
     */
    setLabelExtractor(extractor) {
      this.labelExtractor = extractor;
      return this;
    }

    /**
     * Sets the value extractor function.
     * @param {Function} extractor - The value extractor function
     * @returns {ChartComponent} - The chart component instance for chaining
     */
    setValueExtractor(extractor) {
      this.valueExtractor = extractor;
      return this;
    }

    /**
     * Sets the category extractor function.
     * @param {Function} extractor - The category extractor function
     * @returns {ChartComponent} - The chart component instance for chaining
     */
    setCategoryExtractor(extractor) {
      this.categoryExtractor = extractor;
      return this;
    }

    /**
     * Updates the chart from data state.
     * @param {Array} data - The data array
     * @private
     */
    updateChartFromData(data) {
      if (!data || data.length === 0) {
        return;
      }
      
      // Extract labels and organize data
      const newLabels = [];
      const categoryData = {};
      
      if (this.categoryField) {
        // For grouped data (multiple datasets)
        for (const item of data) {
          const label = this.labelExtractor(item);
          const category = this.categoryExtractor(item);
          const value = this.valueExtractor(item);
          
          if (!newLabels.includes(label)) {
            newLabels.push(label);
          }
          
          if (!categoryData[category]) {
            categoryData[category] = {};
          }
          
          categoryData[category][label] = value;
        }
        
        // Clear existing datasets
        this.datasets = [];
        
        // Create a dataset for each category
        const colors = [
          '#4285f4', '#ea4335', '#fbbc05', '#34a853', 
          '#673ab7', '#3f51b5', '#2196f3', '#03a9f4', 
          '#00bcd4', '#009688', '#4caf50', '#8bc34a'
        ];
        
        let colorIndex = 0;
        for (const category in categoryData) {
          const values = categoryData[category];
          const dataPoints = [];
          
          for (const label of newLabels) {
            dataPoints.push(values[label] || 0);
          }
          
          const color = colors[colorIndex % colors.length];
          this.addDataset(category, dataPoints, color + '33', color);
          colorIndex++;
        }
      } else {
        // For simple data (single dataset)
        const values = [];
        
        for (const item of data) {
          const label = this.labelExtractor(item);
          const value = this.valueExtractor(item);
          
          newLabels.push(label);
          values.push(value);
        }
        
        // Clear existing datasets
        this.datasets = [];
        
        // Add a single dataset
        const backgroundColors = [];
        const borderColors = [];
        
        for (let i = 0; i < values.length; i++) {
          const hue = (i * 30) % 360;
          backgroundColors.push(`hsla(${hue}, 70%, 60%, 0.2)`);
          borderColors.push(`hsla(${hue}, 70%, 60%, 1)`);
        }
        
        this.addDataset(this.valueField, values, backgroundColors, borderColors);
      }
      
      // Update labels
      this.labels = newLabels;
      
      // Update the chart
      if (this.chart) {
        this.updateChart();
      }
    }

    /**
     * Creates the chart.
     * @private
     */
    createChart() {
      if (!this.element || !window.Chart) {
        return;
      }
      
      const ctx = this.element.getContext('2d');
      
      // If there's an existing chart, destroy it first
      if (this.chart) {
        this.chart.destroy();
        this.chart = null;
      }
      
      // Prepare chart configuration
      const config = {
        type: this.type,
        data: {
          labels: this.labels,
          datasets: this.datasets
        },
        options: {
          responsive: true,
          maintainAspectRatio: false,
          ...this.options
        }
      };
      
      // Create the chart
      this.chart = new Chart(ctx, config);
    }

    /**
     * Updates the chart.
     * @private
     */
    updateChart() {
      if (!this.chart) {
        this.createChart();
        return;
      }
      
      this.chart.data.labels = this.labels;
      this.chart.data.datasets = this.datasets;
      this.chart.update();
    }

    /**
     * Renders the component to the DOM.
     * @param {HTMLElement} parent - The parent element to render into
     * @returns {HTMLElement} - The rendered element
     */
    render(parent) {
      const element = super.render(parent);
      
      // Initialize the chart when the canvas is rendered
      setTimeout(() => {
        this.createChart();
      }, 0);
      
      return element;
    }
  }

  // Export the ChartComponent class
  window.jaffolding = window.jaffolding || {};
  window.jaffolding.ChartComponent = ChartComponent;
})();